java/src/main/native/include/jni_utils.hpp

```diff
@@ -0,0 +1,1067 @@
+
+  T get(int index) const {
+    if (orig == NULL) {
+      throw_java_exception(env, "java/lang/NullPointerException", "jobjectArray pointer is NULL");
+    }
+    T ret = static_cast<T>(env->GetObjectArrayElement(orig, index));
+    checkJavaException(env);
+    return ret;
+  }
+
+  void set(int index, const T &val) {
+    if (orig == NULL) {
+      throw_java_exception(env, "java/lang/NullPointerException", "jobjectArray pointer is NULL");
+    }
+    env->SetObjectArrayElement(orig, index, val);
+    checkJavaException(env);
+  }
+};
+
+/**
+ * @brief jobjectArray wrapper to make accessing strings safe through RAII
+ * and convenient.
+ */
+class native_jstringArray {
+private:
+  JNIEnv *const env;
+  native_jobjectArray<jstring> arr;
+  mutable std::vector<native_jstring> cache;
+  mutable std::vector<const char *> c_cache;
+
+  void init_cache() const {
+    if (!arr.is_null() && cache.empty()) {
+      int size = this->size();
+      cache.reserve(size);
+      for (int i = 0; i < size; i++) {
+        cache.push_back(native_jstring(env, arr.get(i)));
+      }
+    }
+  }
+
+  void init_c_cache() const {
+    if (!arr.is_null() && c_cache.empty()) {
+      init_cache();
+      int size = this->size();
+      c_cache.reserve(size);
+      for (int i = 0; i < size; i++) {
+        c_cache.push_back(cache[i].get());
+      }
+    }
+  }
+
+  void update_caches(int index, jstring val) {
+    if (!cache.empty()) {
+      cache[index] = native_jstring(env, val);
+      if (!c_cache.empty()) {
+        c_cache[index] = cache[index].get();
+      }
+    } else if (!c_cache.empty()) {
+      // Illegal state
+      throw std::logic_error("CACHING IS MESSED UP");
+    }
+  }
+
+public:
+  native_jstringArray(JNIEnv *const env, jobjectArray orig) : env(env), arr(env, orig) {}
+
+  bool is_null() const noexcept { return arr.is_null(); }
+
+  int size() const noexcept { return arr.size(); }
+
+  native_jstring &operator[](int index) const { return get(index); }
+
+  native_jstring &get(int index) const {
+    if (arr.is_null()) {
+      throw_java_exception(env, "java/lang/NullPointerException", "jstringArray pointer is NULL");
+    }
+    init_cache();
+    return cache[index];
+  }
+
+  const char **const as_c_array() const {
+    init_c_cache();
+    return c_cache.data();
+  }
+
+  void set(int index, jstring val) {
+    arr.set(index, val);
+    update_caches(index, val);
+  }
+
+  void set(int index, const native_jstring &val) {
+    arr.set(index, val.get_jstring());
+    update_caches(index, val.get_jstring());
+  }
+
+  void set(int index, const char *val) {
+    jstring str = env->NewStringUTF(val);
+    checkJavaException(env);
+    arr.set(index, str);
+    update_caches(index, str);
+  }
+};
+
+/**
+ * Wrapper for gdf_column. This class owns the underlying gdf_column and
+ * release() should be called before assuming ownership of the underlying
+ * gdf_column.
+ */
+class gdf_column_wrapper {
```

Comment:
Will this go away once we have `class cudf::column`?

Reply:
Yes.  I sure hope so.